### PR TITLE
Issue 6867

### PR DESCRIPTION
**Pull request by Janpot**
**What kind of change does this PR introduce?**

Bug fix for https://github.com/webpack/webpack/issues/6867

**Did you add tests for your changes?**

Yes, always

**If relevant, link to documentation update:**

**Summary**
> Explain the **motivation** for making this change. What existing problem does the pull request solve?

Basically, I tried to compile a project with webpack 4, it didn't work.

https://github.com/webpack/webpack/issues/6867 contains more info on what's going on exactly.

**Does this PR introduce a breaking change?**

Only if you rely on it throwing errors when it shouldn't :)

**Other information**

Don't merge yet, I just realized 
```js
export default (class A {

});
```
suffers from the same problem, and some other code probably too.

### CODE REVIEW

File: lib/Parser.js [hunk 1]
```diff
@@ -1178,7 +1178,10 @@ class Parser extends Tapable {
 
 	walkExportDefaultDeclaration(statement) {
 		this.hooks.export.call(statement);
-		if (statement.declaration.id) {
+		if (
+			statement.declaration.id &&
+			statement.declaration.type !== "FunctionExpression"
+		) {
```
Comment (Kovensky) [inline lib/Parser.js]: It's possible that this was an attempt to workaround an old acorn bug where it'd incorrectly parse `export default function () {}` as a function expression.

File: test/configCases/parsing/issue-6867/test.js [hunk 2]
```diff
@@ -0,0 +1,3 @@
+export default (function hello () {
+	return 1;
+})
```
Comment (Kovensky) [inline test/configCases/parsing/issue-6867/test.js]: I'd feel better if there were tests for each permutation.

```js
// function _declaration_, id is null, in native implementations the function is named "default"
export default function() {}
```

```js
// function _expression_, id is null, in native implementations the function is named "default"
export default (function() {})
```

```js
// function _declaration_, id is "foo"
export default function foo() {
  foo = function() {} // this not working is currently a bug in webpack 😝
})
```

```js
// function _expression_
export default (function bar() {})
```

Comment (Kovensky) [reply test/configCases/parsing/issue-6867/test.js]: These same permutations also apply to `class` declarations, but we probably don't need to test them here 🤔

File: lib/Parser.js (hunk 1 above)
Comment (ooflorent) [reply lib/Parser.js]: In acorn 4, `export default function() {` was parsed as a `FunctionExpression`. 
Since acorn 5 it is a `FunctionDeclaration`.

**Comment by Janpot** [comment]
@Kovensky @sokra 
```js
export default (class A {

});
```
is breaking too, how do you feel about 
```js
		if (
			statement.declaration.id &&
			!statement.declaration.type.endsWith("Expression")
		)
```
? is that too hacky?

**Comment by ooflorent** [comment]
@Janpot yes it is. You could test both node types.

**Comment by Janpot** [comment]
@Kovensky @sokra @ooflorent I think I addressed your comments. 

@Kovensky Not sure why you mention `// this not working is currently a bug in webpack 😝`. Do you think it makes sense to address that issue as part of this PR?

**Comment by Kovensky** [comment]
Nah, it's definitely a separate issue. Webpack doesn't actually detect when an export is mutable (mutated) or immutable, it just guesses that `function` and `const` are immutable and everything else is mutable.

**Comment by Janpot** [comment]
Could it be that the coverage went down, just because the amount of lines increased more than the amount of statements?

File: test/configCases/parsing/issue-6867/index.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
+it("should compile default export unnamed function declaration", function() {
```
Comment (sokra) [inline test/configCases/parsing/issue-6867/index.js]: Could you move this test case from `configCases` to `cases`?

Comment (Janpot) [reply test/configCases/parsing/issue-6867/index.js]: done

File: test/cases/parsing/issue-6867/index.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+});
+
+it("should compile default export named function declaration", function() {
+	return import(/* webpackChunkName: "g" */ "./g")
+		.then(({ default: G }) => {
+			new G()
+		});
+});
+
+it("should compile default export named function expression", function() {
```
Comment (Kovensky) [inline test/cases/parsing/issue-6867/index.js]: `class expression`; same for lines 30, 38 and 45

**Comment by sokra** [comment]
Thanks

**Pull request merged**